src/boot/efi/util.c

```diff
@@ -697,3 +697,92 @@ char16_t *get_extra_dir(const EFI_DEVICE_PATH *file_path) {
         remove_boot_count(file_path_str);
         return xasprintf("%ls.extra.d", file_path_str);
 }
+
+/*
+ * The EFI_LOAD_OPTION descriptor has the following layout:
+ *        u32 Attributes;
+ *        u16 FilePathListLength;
+ *        u16 Description[];
+ *        efi_device_path_protocol_t FilePathList[];
+ *        u8 OptionalData[];
+ *
+ * This function validates and unpacks the variable-size data fields.
+ */
+static bool efi_load_option_unpack(
+                EFI_LOAD_OPTION_UNPACKED *dest,
+                const EFI_LOAD_OPTION *src, size_t size) {
```

Comment:
if this is copied over from Linux sources, then we should make that clear in a comment, and we should explain why that is OK license-wise.

linux is gpl2.0-only, we are lgpl-2.1+

i think we should have no problem with linking this together, but our spdx header here says lgpl, hence let's maybe move this helper into a separate .c/.h file pair, with a more correct spdx header.

Reply:
uuh actually that's not fine - if it's a verbatim copy and paste, it will effectively mean systemd-boot and systemd-stub change license from LGPL-2.1+ to GPL-2.0-only, and that 'only' is problematic. Given it's a change of license in a more restrictive way, we'd need an ACK from every contributor we ever had.

Reply:
that's not true. our code remains lgpl. it's that one file that would be gpl2. the fact that we link this together has zero effect on the license of our own code. we just need to keep it reasonably separate, that's all.

of course, the combined binary would distributable only under the provisions of gpl2 (because it is the more restrictive) one.

we don#t need to ask anyone about this at all. we do not relicense anything.

Reply:
By systemd-boot/stub I meant the distributed binaries of course - it matters little what individual files are, for distribution purposes, the aggregated license is what matters. And I'm pretty sure we do not want to change the license of the binaries like that. I did a lot of work to clean things up a couple of years ago, see LICENSES/ subdir and the readme:

> New sources that cannot be distributed under LGPL-2.1-or-later will no longer
be accepted for inclusion in the systemd project to maintain license uniformity.